src/cargo/core/compiler/job_queue.rs

```diff
@@ -854,34 +872,43 @@ impl<'cfg> DrainState<'cfg> {
             // we'll just naturally abort the compilation operation but for 1
             // we need to make sure that the metadata is flagged as produced so
             // send a synthetic message here.
-            if state.rmeta_required.get() && sender.result.is_ok() {
-                messages.push(Message::Finish(id, Artifact::Metadata, Ok(())));
-            }
-
-            // Use a helper struct with a `Drop` implementation to guarantee
-            // that a `Finish` message is sent even if our job panics. We
-            // shouldn't panic unless there's a bug in Cargo, so we just need
-            // to make sure nothing hangs by accident.
-            struct FinishOnDrop<'a> {
-                messages: &'a Queue<Message>,
-                id: JobId,
-                result: CargoResult<()>,
+            if state.rmeta_required.get() && result.is_ok() {
+                state
+                    .messages
+                    .push(Message::Finish(id, Artifact::Metadata, Ok(())));
             }
 
-            impl Drop for FinishOnDrop<'_> {
-                fn drop(&mut self) {
-                    let msg = mem::replace(&mut self.result, Ok(()));
-                    self.messages
-                        .push(Message::Finish(self.id, Artifact::All, msg));
-                }
-            }
+            state
+                .messages
+                .push(Message::Finish(id, Artifact::All, result));
         };
 
         match fresh {
-            Freshness::Fresh => self.timings.add_fresh(),
-            Freshness::Dirty => self.timings.add_dirty(),
+            Freshness::Fresh => {
+                self.timings.add_fresh();
+                // Running a fresh job on the same thread is often much faster than spawning a new
+                // thread to run the job.
+                doit(JobState {
+                    id,
+                    messages: messages.clone(),
+                    output: Some(cx.bcx.config),
+                    rmeta_required: Cell::new(rmeta_required),
+                    _marker: marker::PhantomData,
+                });
+            }
+            Freshness::Dirty => {
+                self.timings.add_dirty();
+                scope.spawn(move |_| {
+                    doit(JobState {
+                        id,
+                        messages: messages.clone(),
+                        output: None,
+                        rmeta_required: Cell::new(rmeta_required),
+                        _marker: marker::PhantomData,
+                    })
+                });
+            }
         }
-        scope.spawn(move |_| doit());
```

Comment:
There was an amount of back and forth about this [awhile ago](https://github.com/rust-lang/cargo/pull/7838). Originally Cargo spawned threads for everything, but then I eventually optimized it to do what you have here. In #7838 we reverted back to what we have today, but IIRC @ehuss did some measurements and found the cost to be negligible. I can't seem to find a record of that conversation though, @ehuss do you remember that?

Reply:
From the PR description of #7838:

> On Linux and Windows, it was generally about 0 to 5% slower.

This roughly matches my results.

This PR keeps not buffering the rustc output, but for fresh jobs it directly sends it to the terminal without the indirection through the message queue, thereby slightly improving performance.

Reply:
One issue I believe with this PR is that it can still deadlock. If the message queue is full and a fresh job tries to push onto it then the deadlock happens. That won't happen for stdout but other messages go through the queue as well (like build plans and such).

Another issue which I'm having trouble remembering is that we want the main "event loop" to complete quickly each iteration, and if we're copying a lot of rustc output onto the console that isn't happening. I forget the consequences of a slow loop iteration though.

Reply:
The other messages don't use the bounded `push_bounded`, but the unbounded `push`.

Reply:
> did some measurements and found the cost to be negligible. I can't seem to find a record of that conversation though, @ehuss do you remember that?

Just the note in that PR that it was expected to cost about 5%.  That number will vary significantly based on the project and the system.  I ran some tests with this PR, and it is pretty much the same (no difference on macos, about 5% on Linux).

My main concern from back then with this approach is that it introduces some subtle complexity that didn't really seem to be worth the 5% improvement, but I can't think of any specific problems with this PR other than being harder to understand.

Reply:
Ah sorry I forgot about `push_bounded` vs `push`, makes sense to me!

@ehuss I'm curious, how are you measuring? I'd expect that thread creation on Windows and macOS to be a good deal more expensive than on Linux

Reply:
Just using hyperfine, like this:

```
hyperfine -w1  "~/Proj/cargo/target/release/cargo check" "~/Proj/cargo2/target/release/cargo check"
```

In various projects of various sizes.  Larger projects should show a bigger difference.

I just re-ran the test on a larger project (libra, 600+ packages), and the numbers look comparable to linux, so I was probably just comparing a smaller project where the difference was less noticeable.

```
master:
  Time (mean ± σ):     851.4 ms ±  23.5 ms    [User: 574.5 ms, System: 247.4 ms]
  Range (min … max):   820.7 ms … 897.5 ms    10 runs

improve_perf:
  Time (mean ± σ):     815.7 ms ±  34.4 ms    [User: 567.6 ms, System: 289.2 ms]
  Range (min … max):   757.6 ms … 857.3 ms    10 runs

Summary
  '~/Proj/rust/cargo3/target/release/cargo check' ran
    1.04 ± 0.05 times faster than '~/Proj/rust/cargo2/target/release/cargo check'
```